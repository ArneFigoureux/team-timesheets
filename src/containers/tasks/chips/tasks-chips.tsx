import React from "react";
import { Chip, ChipSet, ChipSetProps } from "@rmwc/chip";
import { Icon } from "@rmwc/icon";

import { ITask } from "../../../../common";

export const TasksChips = ({
    onTaskInteraction,
    selectedTaskIds,
    tasks,
    filter,
    choice,
    ...chipSetProps
}: {
    onTaskInteraction: (id: string, selected?: boolean) => void;
    selectedTaskIds?: string[];
    tasks: (ITask & { id: string })[];
} & ChipSetProps) => {
    const chips = (
        <>
            {
                tasks.map((t) => {
                    const { id: taskId, name: taskName, icon: taskIcon = undefined } = t;
                    const leadingIcon = taskIcon
                        ? <Icon icon={taskIcon} />
                        : undefined;

                        const selected = selectedTaskIds?.indexOf(taskId) !== -1;

                    return (
                        <Chip
                            checkmark={filter}
                            icon={leadingIcon}
                            selected={selected}
                            onInteraction={() => onTaskInteraction(taskId, selected!)}
                            id={taskId}
                            label={taskName!}
                            key={taskId}
                        />
                    );
                })
            }
        </>
    );

    return (
        <ChipSet
            filter={filter}
            choice={choice}
            {...chipSetProps}
        >
            {chips}
        </ChipSet>
    );
};