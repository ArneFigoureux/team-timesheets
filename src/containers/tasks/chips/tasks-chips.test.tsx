import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { TasksChips } from "./tasks-chips";

const tasks = [
    {
        id: "1",
        name: "Task 1",
        icon: "people",
    },
    {
        id: "2",
        name: "Task 2",
    },
    {
        id: "3",
        name: "Task 3",
        icon: "people",
    },
];

describe("TasksChips", () => {
    it("should render empty when no tasks", () => {
        const { asFragment } = render(<TasksChips
            onTaskInteraction={jest.fn()}
            tasks={[]}
        />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render chip for each task", () => {
        const { getByText } = render(<TasksChips
            onTaskInteraction={jest.fn()}
            tasks={tasks}
        />);

        expect(getByText("Task 1"));
        expect(getByText("Task 2"));
        expect(getByText("Task 3"));
    });

    it("should show selected task ids", () => {
        const { container } = render(<TasksChips
            onTaskInteraction={jest.fn()}
            tasks={tasks}
            selectedTaskIds={["3"]}
        />);

        expect(container.querySelector(".mdc-chip--selected")).not.toBeNull();
    });

    it("should call onInteraction when chips are being interacted with", () => {
        const onInteraction = jest.fn();

        const { container } = render(<TasksChips
            onTaskInteraction={onInteraction}
            tasks={tasks}
            selectedTaskIds={["3"]}
        />);

        const selectedChip = container.querySelector(".mdc-chip--selected");
        expect(selectedChip).not.toBeNull();

        fireEvent.click(selectedChip!);

        expect(onInteraction).toBeCalledTimes(1);
    });
});

