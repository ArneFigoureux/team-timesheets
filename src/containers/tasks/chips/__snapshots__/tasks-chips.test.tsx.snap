// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksChips should render empty when no tasks 1`] = `
<DocumentFragment>
  <div
    class="mdc-chip-set"
  />
</DocumentFragment>
`;
